### PR TITLE
Fix Unwraps, Add Flexi Logger for Log Rotation, and Update Examples

### PR DESCRIPTION
**Pull request by vsawant10949**
This pull request addresses several issues within the code base, focusing on fixing unwraps, adding log rotation with Flexi logger, and updating examples to ensure they work correctly in the designated directory.

Changes Made:
- Fixed all occurrences of unwraps in the code base, replacing them with appropriate error handling to improve code robustness and prevent potential panics.
- Implemented Flexi logger for log rotation. Log files will now be retained for 7 days, and compressed zip files will be stored for 2 days, allowing for efficient log management.
- Updated examples to work seamlessly in the specified working directory, ensuring that they reflect accurate usage patterns and provide a better understanding of the code.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -35,7 +35,7 @@ fn main() {
     CombinedLogger::init(vec![
         TermLogger::new(LevelFilter::Info, logger_config).unwrap()
     ])
-    .unwrap();
+        .unwrap();
```
Comment (nopestack) [inline src/main.rs:38]: unhandled unwrap here

File: src/main.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ fn main() {
...
 
     let mut node_map = HashMap::new();
     let mut id: u32 = 0;
```
Comment (nopestack) [inline src/main.rs:41]: There's several unhandled unwraps in the code below

File: src/node/mod.rs [hunk 3]
```diff
@@ -21,7 +23,25 @@ use std::sync::mpsc::{channel, Receiver, Sender};
 use std::sync::{Arc, Mutex};
 use std::thread;
 use std::time::Duration;
+lazy_static! {
+    static ref LOGGER_INIT: std::sync::Once = std::sync::Once::new();
+}
 
+fn initialize_logger() {
```
Comment (nopestack) [inline src/node/mod.rs:30]: Remove this logger initialization altogether and instead rely on [tracing](https://github.com/tokio-rs/tracing) which is the standard tool we use for instrumenting our codebases.

File: src/node/mod.rs [hunk 4]
```diff
@@ -43,10 +63,14 @@ impl Node {
         bootstrap: Option<NodeData>,
         // request_timeout: Option<u64>,
     ) -> Self {
+        initialize_logger();
```
Comment (nopestack) [inline src/node/mod.rs:66]: Remove this call.

File: src/node/mod.rs [hunk 5]
```diff
@@ -43,10 +63,14 @@ impl Node {
...
         let socket = UdpSocket::bind(addr).expect("Error: could not bind to address.");
+        let addr = socket
+            .local_addr()
```
Comment (nopestack) [inline src/node/mod.rs:69]: Handle this unwrap properly such that this function does not trigger a panic.

File: src/node/mod.rs [hunk 6]
```diff
@@ -76,6 +100,9 @@ impl Node {
         ret
     }
 
+    pub fn routing_table(&self)->RoutingTable{
+        self.routing_table.lock().unwrap().clone()
```
Comment (nopestack) [inline src/node/mod.rs:104]: unhandled unwrap here

File: src/node/mod.rs [hunk 7]
```diff
@@ -176,9 +209,11 @@ impl Node {
         let target_key = self.node_data.id;
 
         self.lookup_nodes(&target_key, true);
-
-        let bucket_size = { self.routing_table.lock().unwrap().size() };
-
+        let bucket_size = if let Ok(routing_table) = self.routing_table.lock() {
```
Comment (nopestack) [inline src/node/mod.rs:212]: a call to an `unwrap_or` variant may be a better fit here

File: src/node/mod.rs [hunk 8]
```diff
@@ -223,28 +258,55 @@ impl Node {
         );
         self.clone().update_routing_table(request.sender.clone());
         let receiver = (*self.node_data).clone();
+        const RETRY_ATTEMPTS: i32 = 3;
```
Comment (nopestack) [inline src/node/mod.rs:261]: define this constant elsewhere and document it so its not so easy to miss.

**Review by nopestack — CHANGES_REQUESTED** [review]
Please address the comments left on this PR

**Comment by nopestack** [comment]
Marked this as draft to prevent accidental merges while the log comments are addressed.

File: src/main.rs (hunk 1 above)
Comment (vsawant10949) [reply src/main.rs:38]: Yes

File: src/protocol.rs [hunk 9]
```diff
@@ -50,6 +50,7 @@ pub enum ResponsePayload {
     Nodes(Vec<NodeData>),
     Value(String),
     Pong,
+    Error(String)
```
[the remark this answers is not in the document]
Comment (vsawant10949) [reply src/protocol.rs:53]: In case if locking fails, the functions needs a response payload.

File: src/protocol.rs [hunk 10]
```diff
@@ -76,12 +77,13 @@ impl Protocol {
...
             loop {
-                let (len, _src_addr) = protocol.socket.recv_from(&mut buffer).unwrap();
-                let message = bincode::deserialize(&buffer[..len]).unwrap();
-
-                if tx.send(message).is_err() {
-                    warn!("Protocol: Connection closed.");
-                    break;
+                if let Ok( (len, _src_addr) )= protocol.socket.recv_from(&mut buffer){
+                   if let Ok(message) =  bincode::deserialize(&buffer[..len]){
+                        if tx.send(message).is_err() {
```
Comment (nopestack) [inline src/protocol.rs:82]: log this error instead of checking that there is an error

File: Cargo.toml [hunk 11]
```diff
@@ -21,7 +21,8 @@ simplelog = "0.5"
 time = "0.1"
 clap = {version="4.0.32",features=["derive"]}
 hex="*"
-
+lazy_static = "1.4.0"
```
Comment (nopestack) [inline Cargo.toml:24]: Please remove these dependencies once their usage is refactored away.

File: src/node/mod.rs (hunk 3 above)
Comment (vsawant10949) [reply src/node/mod.rs:30]: Done

File: src/node/mod.rs (hunk 8 above)
Comment (vsawant10949) [reply src/node/mod.rs:261]: Done

File: src/node/mod.rs (hunk 7 above)
Comment (vsawant10949) [reply src/node/mod.rs:212]: Done

File: src/node/mod.rs [hunk 12]
```diff
@@ -96,13 +123,18 @@ impl Node {
         let mut node = self.clone();
 
         thread::spawn(move || loop {
-            let routing_table = node.routing_table.lock().unwrap().clone();
+            let routing_table = match node.routing_table.lock() {
+                Ok(table) => table.clone(),
+                Err(_) => {
+                    continue;
```
[the remark this answers is not in the document]
Comment (vsawant10949) [reply src/node/mod.rs:129]: Done

File: src/node/mod.rs (hunk 5 above)
Comment (vsawant10949) [reply src/node/mod.rs:69]: Done

File: src/node/mod.rs (hunk 6 above)
Comment (vsawant10949) [reply src/node/mod.rs:104]: Done

File: Cargo.toml (hunk 11 above)
Comment (vsawant10949) [reply Cargo.toml:24]: Done

File: src/protocol.rs (hunk 10 above)
Comment (vsawant10949) [reply src/protocol.rs:82]: Done

File: Cargo.toml [hunk 13]
```diff
@@ -21,7 +20,7 @@ simplelog = "0.5"
 time = "0.1"
 clap = {version="4.0.32",features=["derive"]}
 hex="*"
-
+tracing = "0.1.36"
```
Comment (nopestack) [inline Cargo.toml:23]: 👍

File: src/key.rs [hunk 14]
```diff
@@ -1,5 +1,4 @@
 use crate::KEY_LENGTH;
-use rand;
```
Comment (nopestack) [inline src/key.rs:2]: Why was this removed?

File: src/lib.rs [hunk 15]
```diff
@@ -1,107 +1,3 @@
-//! # kademlia-dht-rs
```
Comment (nopestack) [inline src/lib.rs:1]: Why where these doc comments removed?

File: src/lib.rs [hunk 16]
```diff
@@ -144,3 +40,6 @@ const SAMPLE_PERCENTAGE_NODES_TO_PING: f64 = 0.3;
 
 /// Ping Nodes after every N seconds
 const PING_TIME_INTERVAL: u64 = 10;
+
+/// `RETRY_ATTEMPTS` is a constant that specifies the maximum number of retry attempts for any operation
+const RETRY_ATTEMPTS: i32 = 3;
```
Comment (nopestack) [inline src/lib.rs:45]: 👍

**Review by nopestack — APPROVED** [review]
LGTM, only a minor issue pointed out.

File: src/node/mod.rs [hunk 17]
```diff
@@ -287,49 +336,55 @@ impl Node {
...
-                // dbg!(
-                //     // "{} - Request to {} timed out after waiting for {} milliseconds",
-                //     &self.node_data.id,
-                //     &self.node_data.addr,
-                //     &dest.addr,
-                //     &REQUEST_TIMEOUT,
-                // );
-
+            Ok(Err(_)) => {
+                println!(
```
Comment (nopestack) [inline src/node/mod.rs:370]: minor: this needs to be removed
